README.md
# Wind Turbines (Redux)

This mod adds multiple tiers of wind turbines that turn with the wind (in up to 32 steps) and will provide power (configurable base). You can choose whenever it should provide power at a fixed rate or vary wind speed. In order to balance them against solars they do have a rather big footprint and need to have some distance betweem them. You can fill that space with other stuff though.

This version is a redux of the original by Ownly Me. In comparison to Wind Turbines 2 by majuss it should properly limit power production and it also features configurable base power from Wind Turbines 2 Configurable Power by ethanatos.

The latter also feature support for WindSpeedChanging. This mod detects its presence and will let it to do its thing if it's detected. If you are playing with varying wind speed enabled and WindSpeedChanging is not detected this mod will vary wind speed on its own.

Copyright wise this mod is CC BY-SA 3.0 DE, except for the [Unity asset that was used for the wind turbine](https://assetstore.unity.com/packages/3d/props/exterior/wind-turbine-26741) itself. The original author is OwnlyMe.

changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.19.03
Date: 2023-02-18
  Bugfixes:
    - Included a migration which should remove duplicate animations by clearing up the old animations.

---------------------------------------------------------------------------------------------------
Version: 0.19.02
Date: 2023-02-18
  Features:
    - Support for space-exploration. Wind turbines are not placeable in space.
    - Better integration with WindSpeedChanging.
  Minor Features:
    - Clear localization, remove the mention of the base power from items for now. This might return at a later date.
    - Included a migration for wind turbines 2 and wind turbines 2 - configurable power which should allow this to be used as a drop in replacement.

---------------------------------------------------------------------------------------------------
Version: 0.19.01
Date: 2023-02-13
  Major Features:
    - Updated for 1.1.
    - Most options are runtime global rather than startup settings.
    - Option to configure the base power of turbines.
    - New option to enabled or disabled the mod internal variation of wind speed.

---------------------------------------------------------------------------------------------------
Version: 0.18.11
  Features:
    - After re-measuring the length of the blades, your engineers have come to the conclusion that it's safe to build the turbines 3 tiles closer to each other.
    - You lost some weight and can now fit between a (squeak-through) solar panel and a wind turbine

---------------------------------------------------------------------------------------------------
Version: 0.17.10
  Features:
    - Slightly de-randomized the wind speeds and removed hard caps

---------------------------------------------------------------------------------------------------
Version: 0.17.9
  Features:
    - Removed 2 debug messages

---------------------------------------------------------------------------------------------------
Version: 0.17.8
  Features:
    - Fixed top and base not blending seamlessly

---------------------------------------------------------------------------------------------------
Version: 0.17.7
  Features:
    - The mod now randomizes the surface's wind speeds between 0.011 and 0.035 (before it was locked on 0.02)
    - The wind speed determines the animation speed and power output (up to 52% more power than declared, but might also produce 52% less power)
    - Added a startup-setting to make the wind- and animation speed not affect power output

---------------------------------------------------------------------------------------------------
Version: 0.17.5
  Features:
    - Fixed an error when a wind turbine got script-removed
    - Improved performance with huge wind parks

---------------------------------------------------------------------------------------------------
Version: 0.17.4
  Features:
    - Hidden some of the entities from the map view

---------------------------------------------------------------------------------------------------
Version: 0.17.3
  Features:
    - Added some safety checks to prevent crashes

---------------------------------------------------------------------------------------------------
Version: 0.17.2
  Features:
    - Made all sub-entities indestructible to prevent crashes with "fire department"

data.lua
require "entities"
local level = 1
data:extend({
	{
		type = "item",
		name = "ownly_wind_turbine_mk" .. level,
		icon = "__windturbines-redux__/graphics/mk" .. level .. "_icon.png",
		icon_size = 144,
		group = "logistics",
		subgroup = "energy",
		order = "d[solar-panel]-a[wind-turbine]",
		place_result = "ownly_wind_turbine_build_mk" .. level,
		stack_size = 20
	},
	{
		type = "recipe",
		name = "ownly_wind_turbine_mk" .. level,
		energy_required = 30,
		enabled = "false",
		ingredients = {
			{ "plastic-bar", 40 },
			{ "steel-plate", 50 },
			{ "copper-cable", 120 },
			{ "electronic-circuit", 25 },
		},
		result = "ownly_wind_turbine_mk" .. level
	},
	{
		type = "technology",
		name = "ownly_wind_turbine_mk" .. level,
		icon = "__windturbines-redux__/graphics/mk" .. level .. "_icon.png",
		icon_size = 144,
		prerequisites = { "logistic-science-pack", "plastics" },
		effects = {
			{
				type = "unlock-recipe",
				recipe = "ownly_wind_turbine_mk" .. level
			}
		},
		unit = {
			count = 200,
			ingredients = { { "automation-science-pack", 1 }, { "logistic-science-pack", 1 } },
			time = 30
		}
	}
})

level = 2
data:extend({
	{
		type = "item",
		name = "ownly_wind_turbine_mk" .. level,
		icon = "__windturbines-redux__/graphics/mk" .. level .. "_icon.png",
		icon_size = 144,
		group = "logistics",
		subgroup = "energy",
		order = "d[solar-panel]-a[wind-turbine]",
		place_result = "ownly_wind_turbine_build_mk" .. level,
		stack_size = 20
	},
	{
		type = "recipe",
		name = "ownly_wind_turbine_mk" .. level,
		energy_required = 30,
		enabled = "false",
		ingredients = {
			{ "ownly_wind_turbine_mk1", 3 },
			{ "electric-engine-unit", 10 },
			{ "advanced-circuit", 25 },
		},
		result = "ownly_wind_turbine_mk" .. level
	},
	{
		type = "technology",
		name = "ownly_wind_turbine_mk" .. level,
		icon = "__windturbines-redux__/graphics/mk" .. level .. "_icon.png",
		icon_size = 144,
		prerequisites = { "ownly_wind_turbine_mk1", "electric-engine", "electric-energy-distribution-1" },
		effects = {
			{
				type = "unlock-recipe",
				recipe = "ownly_wind_turbine_mk" .. level
			}
		},
		unit = {
			count = 350,
			ingredients = { { "automation-science-pack", 1 }, { "logistic-science-pack", 1 }, { "chemical-science-pack", 1 } },
			time = 30
		}
	}
})

level = 3
data:extend({
	{
		type = "item",
		name = "ownly_wind_turbine_mk" .. level,
		icon = "__windturbines-redux__/graphics/mk" .. level .. "_icon.png",
		icon_size = 144,
		group = "logistics",
		subgroup = "energy",
		order = "d[solar-panel]-a[wind-turbine]",
		place_result = "ownly_wind_turbine_build_mk" .. level,
		stack_size = 20
	},
	{
		type = "recipe",
		name = "ownly_wind_turbine_mk" .. level,
		energy_required = 30,
		enabled = "false",
		ingredients = {
			{ "ownly_wind_turbine_mk2", 3 },
			{ "low-density-structure", 50 },
			{ "processing-unit", 25 },
		},
		result = "ownly_wind_turbine_mk" .. level
	},
	{
		type = "technology",
		name = "ownly_wind_turbine_mk" .. level,
		icon = "__windturbines-redux__/graphics/mk" .. level .. "_icon.png",
		icon_size = 144,
		prerequisites = { "ownly_wind_turbine_mk2", "low-density-structure", "utility-science-pack" },
		effects = {
			{
				type = "unlock-recipe",
				recipe = "ownly_wind_turbine_mk" .. level
			}
		},
		unit = {
			count = 500,
			ingredients = { { "automation-science-pack", 1 }, { "logistic-science-pack", 1 }, { "chemical-science-pack", 1 },
				{ "utility-science-pack", 1 } },
			time = 30
		}
	}
})


control.lua
local windSpeedMultiplier = 22

local turbineNames = {
	["ownly_wind_turbine_build_mk1"] = 1,
	["ownly_wind_turbine_build_mk2"] = 2,
	["ownly_wind_turbine_build_mk3"] = 3,
	["ownly_wind_turbine_mk1"] = 1,
	["ownly_wind_turbine_mk2"] = 2,
	["ownly_wind_turbine_mk3"] = 3,
}

local basePower = settings.global["ownly_windturbines_base_power"].value

local function onInit()
	global.turbines = {}
end

local function onConfigurationChanged()
	if settings.global["ownly_windturbines_base_power"].value ~= basePower then
		basePower = settings.global["ownly_windturbines_base_power"].value
		
		for _, turbine in pairs(global.turbines) do
			if turbine.base and turbine.base.valid then
				turbine.base.electric_buffer_size = (2 ^ (turbine.level - 1) * basePower*1000)/60
			end
		end
	end
	
	-- If the setting for locked power is set we need to set each placed turbine to a fixed power output
	if settings.global["ownly_windturbines_locked_power"].value then
		for _, turbine in pairs(global.turbines) do
			if turbine.base and turbine.base.valid then
				turbine.base.power_production = basePower*1000 / 60 * 2 ^ (turbine.level - 1)
			end
		end
	end
end

local function validateTurbine(unit_number)
	local turbine = global.turbines[unit_number]
	
	if not turbine
		or not turbine.base
		or not turbine.base.valid
		or not turbine.collision_box
		or not turbine.collision_box.valid
	then
		return false
	end
	
	return true
end

local function destroyTurbine(unit_number, died)
	if not global.turbines[unit_number] then
		return
	end
	
	if died then
		if global.turbines[unit_number].base and global.turbines[unit_number].base.valid then
			SCRIPT_KILL = true
			global.turbines[unit_number].base.die(global.turbines[unit_number].base.force)
			SCRIPT_KILL = nil
		end
	elseif global.turbines[unit_number].base then
		global.turbines[unit_number].base.destroy()
	end
	
	rendering.destroy(global.turbines[unit_number].turbine)
	rendering.destroy(global.turbines[unit_number].shadow)
	
	if global.turbines[unit_number].collision_box then
		global.turbines[unit_number].collision_box.destroy()
	end
	
	global.turbines[unit_number] = nil
end

local function eventEntityDied(event)
	if SCRIPT_KILL then
		return
	end
	
	if not turbineNames[event.entity.name] or not global.turbines[event.entity.unit_number] then
		return
	end

	destroyTurbine(event.entity.unit_number, true)
end

local function varyWindSpeed(event)
	if game.active_mods["WindSpeedChanging-Tweaked"] or game.active_mods["WindSpeedChanging"] or settings.global["ownly_windturbines_vary_windspeed"].value == false then
		return
	end
	
	for _, surface in pairs(game.surfaces) do
		-- Vary the wind speed and make sure it's never below 0.001.
		-- Substract 0.002 from wind_speed and add 0.4% of a math.random() value to it.
		-- Take 97.5% of this and add 0.00052.
		surface.wind_speed = math.max(0.001, (surface.wind_speed - 0.002 + math.random() * 0.004) * 0.975 + 0.00052)
	end
end

local function updateTurbines (event)
	local wind_orientations = {}
	local wind_speeds = {}
	
	for surface_name, surface in pairs(game.surfaces) do
		wind_orientations[surface_name] = math.floor(surface.wind_orientation * 32 + 0.499)
		if wind_orientations[surface_name] > 31 then
			wind_orientations[surface_name] = 0
		end
		wind_orientations[surface_name] = (wind_orientations[surface_name] + 16) % 31
		wind_speeds[surface_name] = surface.wind_speed
	end
	
	-- run at max once and increase the number of updates with every turbine built
	-- 1: 1/20 + global.fractional_tick => 0.05 + 0; 0.05 + 0.05; 0.1 + 0.05; ...
	-- 6: 6/20 + global.fractional_tick => 0.3 + 0; 0.3 + 0.3; 0.6 + 0.9;
	-- 20: 20/20 + global.fractional_tick => 1 + 0; 1 + 1; 1 + 1
	local runs = math.min(1, table_size(global.turbines) / 20 + (global.fractional_tick or 0))
	global.fractional_tick = runs % 1
	
	for i = 1, runs do
		local remove_entry
		
		if global.iterate and not global.turbines[global.iterate] then
			global.iterate = nil
		end
		
		if global.iterate then
			if not validateTurbine(global.iterate) then
				-- We should remove this turbine as it is not valid
				remove_entry = global.iterate
			else
				-- The turbine is valid so the animation and power generation should be updated
				local turbine = global.turbines[global.iterate]
				
				-- If the wind direction has changed we should update the turbine heading
				if turbine.orientation ~= wind_orientations[turbine.surface] then
					local level = turbine.level
					local orientation = wind_orientations[turbine.surface]
					
					rendering.set_animation(turbine.turbine, "ownly_wind_turbine_mk" .. level .. "_" .. orientation)
					rendering.set_animation(turbine.shadow, "ownly_wind_turbine_shadow_" .. orientation)
					
					turbine.orientation = orientation
				end
				
				local current_wind_speed = wind_speeds[turbine.base.surface.name] * windSpeedMultiplier
				
				-- If the wind speed has changed we should update the animation speed and power generation
				if turbine.last_speed ~= current_wind_speed then
					local expected_frame = (((game.tick * turbine.last_speed) + turbine.last_offset) % 24)
					local new_frame = (((game.tick * current_wind_speed)) % 24)
					local new_offset = (expected_frame - new_frame) % 24
					
					-- game.print("currently at frame: "..expected_frame)
					-- game.print("new frame w/o offset: "..new_frame)
					-- game.print("new offset: "..new_offset)
					
					rendering.set_animation_speed(turbine.turbine, current_wind_speed)
					rendering.set_animation_speed(turbine.shadow, current_wind_speed)
					rendering.set_animation_offset(turbine.shadow, new_offset)
					rendering.set_animation_offset(turbine.turbine, new_offset)
					
					turbine.last_offset = new_offset
					turbine.last_speed = current_wind_speed
					
					-- If we're not in locked power mode update the generated power
					if not settings.global["ownly_windturbines_locked_power"].value then
						if game.active_mods["WindSpeedChanging"] then
							current_wind_speed = current_wind_speed/10
						end
						
						turbine.base.power_production = basePower*1000 / 60 * 2 ^ (turbine.level - 1) * (current_wind_speed / 0.457)
					end
				end
			end
		end
		
		-- Update iterator
		global.iterate = next(global.turbines, global.iterate)
		
		-- Start from the fornt
		if not global.iterate then
			global.iterate = next(global.turbines, global.iterate)
		end
		
		-- If it's an invalid turbine remove it
		if remove_entry then
			destroyTurbine(remove_entry, true)
		end
	end
end

local function entityBuilt(event)
	local entity = event.created_entity or event.entity
	
	if not turbineNames[entity.name] then
		return
	end
	
	local orientation = 16
	local position = entity.position
	local force = entity.force
	local surface = entity.surface
	local level = turbineNames[entity.name]
	entity.destroy()
	
	local new_entity = surface.create_entity {
			name = "ownly_wind_turbine_mk" .. level,
			position = position,
			force = force
		}
	
	local collision_box = surface.create_entity {
			name = "ownly_wind_turbine_collision_box_mk" .. level,
			position = { position.x + 0, position.y - 0 }, force = force
		}
	
	collision_box.destructible = false
	collision_box.minable = false
	
	new_entity.electric_buffer_size = (2 ^ (level - 1) * basePower*1000)/60
	new_entity.power_production = basePower*1000 / 60 * 2 ^ (level - 1)
	
	global.turbines[new_entity.unit_number] = {
		base = new_entity,
		collision_box = collision_box,
		surface = surface.name,
		orientation = 16,
		level = level,
		turbine = rendering.draw_animation {
			animation = "ownly_wind_turbine_mk" .. level .. "_" .. orientation,
			target = { position.x, position.y - 10 },
			surface = surface, animation_speed = surface.wind_speed * windSpeedMultiplier,
			render_layer = "wires-above"
		},
		shadow = rendering.draw_animation {
			animation = "ownly_wind_turbine_shadow_" .. orientation,
			target = { position.x + 6, position.y - 1 },
			surface = surface,
			animation_speed = surface.wind_speed * windSpeedMultiplier
		},
		last_change = game.tick,
		last_speed = surface.wind_speed * windSpeedMultiplier,
		creation_tick = game.tick,
		last_frame_count = 0,
		last_offset = 0,
	}
end

local function entityRemoved(event)
	local entity = event.entity
	
	if not turbineNames[entity.name] or not global.turbines[entity.unit_number] then
		return
	end
	
	destroyTurbine(entity.unit_number)
end

script.on_init(onInit)
script.on_configuration_changed(onConfigurationChanged)

script.on_event(defines.events.on_built_entity, entityBuilt)
script.on_event(defines.events.on_robot_built_entity, entityBuilt)
script.on_event(defines.events.script_raised_revive, entityBuilt)

script.on_event(defines.events.on_player_mined_entity, entityRemoved)
script.on_event(defines.events.on_entity_died, entityRemoved)
script.on_event(defines.events.on_robot_mined_entity, entityRemoved)

script.on_event(defines.events.on_entity_died, eventEntityDied)

script.on_event(defines.events.on_runtime_mod_setting_changed, onConfigurationChanged)

-- every every 5 seconds vary wind speed on each surface
script.on_nth_tick(300, varyWindSpeed)

-- Every 2/60 of a second update the animation and power generation.
script.on_nth_tick(2, updateTurbines)

settings.lua
data:extend(
    {
        {
            setting_type = "startup",
            name = "ownly_windturbines_low_vram",
            
            type = "bool-setting",
            default_value = false,
        },
        {
            setting_type = "runtime-global",
            name = "ownly_windturbines_locked_power",
            
            type = "bool-setting",
            default_value = false,
        },
        {
            setting_type = "runtime-global",
            name = "ownly_windturbines_vary_windspeed",
            
            type = "bool-setting",
            default_value = true,
        },
        {
            setting_type = "runtime-global",
            name = "ownly_windturbines_base_power",
            
            type = "double-setting",
            default_value = 750.0,
            minimum_value = 1.0,
        }
    }
)


migrations/windturbines2-migration.lua
local windSpeedMultiplier = 22

local turbineNames = {
	["ownly_wind_turbine_build_mk1"] = 1,
	["ownly_wind_turbine_build_mk2"] = 2,
	["ownly_wind_turbine_build_mk3"] = 3,
	["ownly_wind_turbine_mk1"] = 1,
	["ownly_wind_turbine_mk2"] = 2,
	["ownly_wind_turbine_mk3"] = 3,
}

local basePower = settings.global["ownly_windturbines_base_power"].value

for _, surface in pairs(game.surfaces) do
    for level = 1, 3 do
        
        local cb = surface.find_entities_filtered{ name = ("ownly_wind_turbine_mk" .. level)}
        
        print("Found " .. #cb .. " number of turbiens of level " .. level .. " on surface " .. surface.name)
        
        for _,turbine in pairs(cb) do
            local entity = turbine
            
            if not turbineNames[entity.name] then
                return
            end
            
            local orientation = 16
            local position = entity.position
            local force = entity.force
            local surface = entity.surface
            local level = turbineNames[entity.name]
            --entity.destroy()
            
            local new_entity = turbine
            
            local collision_box = surface.find_entities_filtered{position = entity.position, type = "simple-entity-with-force"}[1]
            
            collision_box.destructible = false
            collision_box.minable = false
            
            new_entity.electric_buffer_size = (2 ^ (level - 1) * basePower*1000)/60
            new_entity.power_production = basePower*1000 / 60 * 2 ^ (level - 1)
            
            global.turbines[new_entity.unit_number] = {
                base = new_entity,
                collision_box = collision_box,
                surface = surface.name,
                orientation = 16,
                level = level,
                turbine = rendering.draw_animation {
                    animation = "ownly_wind_turbine_mk" .. level .. "_" .. orientation,
                    target = { position.x, position.y - 10 },
                    surface = surface, animation_speed = surface.wind_speed * windSpeedMultiplier,
                    render_layer = "wires-above"
                },
                shadow = rendering.draw_animation {
                    animation = "ownly_wind_turbine_shadow_" .. orientation,
                    target = { position.x + 6, position.y - 1 },
                    surface = surface,
                    animation_speed = surface.wind_speed * windSpeedMultiplier
                },
                last_change = game.tick,
                last_speed = surface.wind_speed * windSpeedMultiplier,
                creation_tick = game.tick,
                last_frame_count = 0,
                last_offset = 0,
            }
            
        end
    end
end

entities.lua
local data_util = nil
if mods["space-exploration"] then
	data_util = require("__space-exploration__/data_util")
end

COLLISION_BOX = { { -1.2, -1.2 }, { 1.2, 1.2 } }
SHADOW_OFFSET = { 9.5, -3.6 }
MAIN_OFFSET = { 0.15, -11 }
BLADES_OFFSET = { 0.15, -1 }
BLADES_SHADOW_OFFSET = { 9.5, 2.4 }
BLADES_SHADOW_OFFSET = { 3.5, -2.6 }

local basePower = 750

-- create 32 rotations for the shadow
for i = 0, 31 do
	local rotation = i
	if settings.startup["ownly_windturbines_low_vram"].value then
		rotation = math.floor(i / 2) * 2
	end
	
	data:extend({
		{
			type = "animation",
			name = "ownly_wind_turbine_shadow_" .. i,
			filename = "__windturbines-redux__/graphics/LD/shadows/" .. math.floor(rotation) .. ".png",
			width = 400,
			height = 200,
			frame_count = 24,
			line_length = 5,
			shift = BLADES_SHADOW_OFFSET,
			animation_speed = 1,
			draw_as_shadow = true,
			scale = 4,
			hr_version =
			{
				filename = "__windturbines-redux__/graphics/HD/shadows/" .. math.floor(rotation) .. ".png",
				width = 800,
				height = 400,
				frame_count = 24,
				line_length = 5,
				shift = BLADES_SHADOW_OFFSET,
				scale = 2,
				animation_speed = 1,
				draw_as_shadow = true,
			}
		}
	})
end

-- create 3 levels of turbines
for level = 1, 3 do
	data:extend({
		{
			type = "simple-entity-with-force",
			name = "ownly_wind_turbine_collision_box_mk" .. level,
			icon = "__base__/graphics/icons/steel-chest.png",
			icon_size = 32,
			flags = { "placeable-neutral", "player-creation", "not-on-map" },
			order = "s-e-w-f",
			max_health = 500,
			corpse = "small-remnants",
			collision_box = { { -16.5, -16.5 }, { 16.5, 16.5 } },
			collision_mask = { "resource-layer" },
			selection_box = { { -0.9, -0.9 }, { 0.9, 0.9 } },
			selectable_in_game = false,
			render_layer = "wires-above",
			random_animation_offset = false,
			picture = {
				filename = "__windturbines-redux__/graphics/transparent32.png",
				width = 32,
				height = 32,
			},
		}
	})
	
	-- create 32 rotations for each wind turbine
	for i = 0, 31 do
		local rota = i
		if settings.startup["ownly_windturbines_low_vram"].value then
			rota = math.floor(i / 2) * 2
		end
		
		data:extend({
			{
				type = "animation",
				name = "ownly_wind_turbine_mk" .. level .. "_" .. i,
				filename = "__windturbines-redux__/graphics/LD/mk" .. level .. "/" .. math.floor(rota) .. ".png",
				width = 300,
				height = 400,
				frame_count = 24,
				line_length = 6,
				shift = BLADES_OFFSET,
				animation_speed = 1,
				scale = 2,
				hr_version = {
					filename = "__windturbines-redux__/graphics/HD/mk" .. level .. "/" .. math.floor(rota) .. ".png",
					width = 600,
					height = 800,
					frame_count = 24,
					line_length = 6,
					shift = BLADES_OFFSET,
					animation_speed = 1,
				}
			},
		})
	end
	
	data:extend({
		{
			type = "electric-energy-interface",
			name = "ownly_wind_turbine_build_mk" .. level,
			icon = "__windturbines-redux__/graphics/mk" .. level .. "_icon.png",
			icon_size = 144,
			order = "a",
			flags = { "placeable-neutral", "player-creation" },
			minable = { mining_time = 0.5, result = "ownly_wind_turbine_mk" .. level },
			max_health = 500,
			corpse = "medium-remnants",
			effectivity = 1,
			resistances = {
				{
					type = "physical",
					percent = 10
				}
			},
			collision_box = COLLISION_BOX,
			selection_box = { { -0.5, -0.5 }, { 0.5, 0.5 } },
			collision_mask = { "resource-layer", "object-layer", "player-layer", "water-tile" },
			energy_source = {
				type = "electric",
				usage_priority = "primary-output",
				input_flow_limit = "0kW",
				render_no_power_icon = false
			},
			energy_production = "0kW",
			energy_usage = "0kW",
			animations = { layers = {
				{
					filename = "__windturbines-redux__/graphics/LD/mk" .. level .. "/build_dummy.png",
					width = 300,
					height = 400,
					frame_count = 24,
					line_length = 6,
					shift = MAIN_OFFSET,
					animation_speed = 0.4,
					scale = 2,
					hr_version = {
						filename = "__windturbines-redux__/graphics/HD/mk" .. level .. "/build_dummy.png",
						width = 600,
						height = 800,
						frame_count = 24,
						line_length = 6,
						shift = MAIN_OFFSET,
						animation_speed = 0.4,
					}
				},
				{
					filename = "__windturbines-redux__/graphics/LD/shadows/16.png",
					width = 400,
					height = 200,
					frame_count = 24,
					line_length = 5,
					shift = SHADOW_OFFSET,
					animation_speed = 0.4,
					draw_as_shadow = true,
					scale = 4,
					hr_version = {
						filename = "__windturbines-redux__/graphics/HD/shadows/16.png",
						width = 800,
						height = 400,
						frame_count = 24,
						line_length = 5,
						scale = 2,
						shift = SHADOW_OFFSET,
						animation_speed = 0.4,
						draw_as_shadow = true,
					}
				}
			} },
			
			random_animation_offset = false,
			working_sound = {
				sound = {
					filename = "__base__/sound/train-wheels.ogg",
					volume = 0.6
				},
				match_speed_to_activity = true,
			},
			min_perceived_performance = 0.25,
		},
		{
			type = "electric-energy-interface",
			name = "ownly_wind_turbine_mk" .. level,
			icon = "__windturbines-redux__/graphics/mk" .. level .. "_icon.png",
			icon_size = 144,
			order = "a",
			flags = { "placeable-neutral", "player-creation" },
			minable = { mining_time = 0.5, result = "ownly_wind_turbine_mk" .. level },
			placeable_by = { item = "ownly_wind_turbine_mk" .. level, count = 1 },
			max_health = 500,
			corpse = "big-remnants",
			effectivity = 1,
			resistances = {
				{
					type = "physical",
					percent = 10
				}
			},
			collision_box = COLLISION_BOX,
			selection_box = { { -0.9, -0.9 }, { 0.9, 0.9 } },
			collision_mask = { "resource-layer", "object-layer", "player-layer", "water-tile" },
			energy_source = {
				type = "electric",
				usage_priority = "primary-output",
				input_flow_limit = "0kW",
				render_no_power_icon = false
			},
			energy_production = "0kW",
			energy_usage = "0kW",
			picture = {
				filename = "__windturbines-redux__/graphics/LD/mk" .. level .. "/base.png",
				width = 300,
				height = 400,
				frame_count = 1,
				line_length = 1,
				shift = MAIN_OFFSET,
				animation_speed = 0.4,
				scale = 2,
				hr_version = {
					filename = "__windturbines-redux__/graphics/HD/mk" .. level .. "/base.png",
					width = 600,
					height = 800,
					frame_count = 1,
					line_length = 1,
					shift = MAIN_OFFSET,
					animation_speed = 0.4,
				}
			},
			
			random_animation_offset = false,
			working_sound = {
				sound = {
					filename = "__base__/sound/train-wheels.ogg",
					volume = 0.6
				},
				match_speed_to_activity = true,
			},
			min_perceived_performance = 0.25,
		},
	})
	
	if mods["space-exploration"] then
		table.insert(data.raw["electric-energy-interface"][( "ownly_wind_turbine_build_mk" .. level )].collision_mask, space_collision_layer)
		data_util.collision_description(data.raw["electric-energy-interface"][( "ownly_wind_turbine_build_mk" .. level )])
		
		table.insert(data.raw["electric-energy-interface"][( "ownly_wind_turbine_mk" .. level )].collision_mask, space_collision_layer)
		data_util.collision_description(data.raw["electric-energy-interface"][( "ownly_wind_turbine_mk" .. level )])
	end
end